src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs

```diff
@@ -99,22 +99,73 @@ internal override bool IsInsertionTrigger(SourceText text, int characterPosition
             return ValueTuple.Create(displayText, insertionText);
         }
 
-        private static CompletionItemRules s_importDirectiveRules =
-            CompletionItemRules.Create(commitCharacterRules: ImmutableArray.Create(CharacterSetModificationRule.Create(CharacterSetModificationKind.Replace, '.', ';')));
-        private static CompletionItemRules s_importDirectiveRules_preselect = s_importDirectiveRules.WithSelectionBehavior(CompletionItemSelectionBehavior.HardSelection);
+        protected override CompletionItemRules GetCompletionItemRules(List<ISymbol> symbols, SyntaxContext context, bool preselect)
+        {
+            return GetCachedRule(context.IsInImportsDirective, preselect, ((CSharpSyntaxContext)context).IsPossibleTupleContext);
+        }
 
-        // '<' should not filter the completion list, even though it's in generic items like IList<>
-        private static readonly CompletionItemRules s_itemRules = CompletionItemRules.Default.
-            WithFilterCharacterRule(CharacterSetModificationRule.Create(CharacterSetModificationKind.Remove, '<')).
-            WithCommitCharacterRule(CharacterSetModificationRule.Create(CharacterSetModificationKind.Add, '<'));
+        // matrix of rules depending on contexts, with the following dimensions:
```

Comment:
It would be *much* cleaner to just have a dictionary and key off of ... wait for it... a Tuple :)